### ISSUE TITLE
Conflict with abbreviations, toc extension, and typeset plugin

### ISSUE THREAD

**Issue by jonaharagon** [squidfunk/mkdocs-material#5579]
### Context

Updating to the latest mkdocs-material-insiders breaks some tables of contents on our staging site: https://deploy-preview-2177.preview.privacyguides.dev/en/vpn/#ipv6-support

### Bug description

In the Table of Contents with the typeset plugin enabled, if there's an abbreviation in a heading, the abbreviation and any subsequent text doesn't get rendered at all.

If the abbreviation is at the very beginning of the header it does show the entire heading text, but the permalink icon gets erroneously included at the end:

<img width="1423" alt="NK3dksoRsmGJsQD2" src="https://github.com/squidfunk/mkdocs-material/assets/3637842/777c25fd-d2f0-41ef-91e0-5040704b3d16">

### Related links

- [Reporting a bug](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/)
- https://squidfunk.github.io/mkdocs-material/setup/extensions/python-markdown/?h=toc#table-of-contents
- https://squidfunk.github.io/mkdocs-material/setup/extensions/python-markdown/?h=abbr#abbreviations
- https://squidfunk.github.io/mkdocs-material/reference/?h=typeset#built-in-typeset-plugin
- https://github.com/squidfunk/mkdocs-material/issues/5165
- https://github.com/squidfunk/mkdocs-material/discussions/5529

### Reproduction

[example.zip](https://github.com/squidfunk/mkdocs-material/files/11607225/example.zip)

### Steps to reproduce

1. Enable typeset plugin
2. Enable TOC
3. Enable abbreviations
4. Use abbreviation in heading

### Before submitting

- [X] I have read and followed the [bug reporting guidelines](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/).
- [X] I have attached links to [the documentation](https://squidfunk.github.io/mkdocs-material), and possibly related [issues](https://github.com/squidfunk/mkdocs-material/issues) and [discussions](https://github.com/squidfunk/mkdocs-material/discussions).
- [X] I assure that I have [removed all customizations](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/#remove-customizations) before submitting this bug report.
- [X] I have attached a __.zip file__ with a [minimal reproduction](https://squidfunk.github.io/mkdocs-material/guides/creating-a-reproduction/).

**Comment by squidfunk** [comment]
Thanks for reporting. So just to be clear – it works if you downgrade to [9.1.14+insiders-4.35.1](https://github.com/squidfunk/mkdocs-material-insiders/releases/tag/9.1.14-insiders-4.35.1), right? If so, it is very likely related to the change added in #5536. /cc @pawamoy.

**Comment by pawamoy** [comment]
Thanks for the ping @squidfunk  :slightly_smiling_face: 
I tried locally, reverting the change from #5536, and the issue persists.

**Comment by pawamoy** [comment]
I think `typeset`'s regular expression for finding headings is responsible here.

Here is an example heading containing the abbreviation, from OP's repro:

```html
<h2 id="othertext-abbreviation-moretext">
  OtherText <abbr title="Test abbreviation!">Abbreviation</abbr> MoreText
  <a class="headerlink" href="#othertext-abbreviation-moretext" title="Permanent link">¶</a>
</h2>
```

It seems the non-greedy operator does not behave as expected :thinking: 
And even if it did, you'd lose the text in the `<abbr>` tag :thinking: 

```pycon
>>> import re
>>> html = """<h2 id="othertext-abbreviation-moretext">OtherText <abbr title="Test abbreviation!">Abbreviation</abbr> MoreText<a class="headerlink" href="#othertext-abbreviation-moretext" title="Permanent link">¶</a></h2>"""
re.findall(r"id=\"([^\"]+).*?>(.*?)<\/h(\d)", html, re.I)
[('othertext-abbreviation-moretext',
  'OtherText <abbr title="Test abbreviation!">Abbreviation</abbr> MoreText<a class="headerlink" href="#othertext-abbreviation-moretext" title="Permanent link">¶</a>',
  '2')]
```

**Comment by squidfunk** [comment]
Ah, interesting, thanks for debugging! I'll check that asap!

**Comment by squidfunk** [comment]
Fixed in [`72c3d0d5f`](https://github.com/squidfunk/mkdocs-material-insiders/commit/72c3d0d5feee9c885d6eee9fb859aab3616aed2b) – I adapted the fix proposed by @pawamoy and made it a little more explicit. Thanks again for the excellent reproduction, that made fixing and checking dead simple. Perfect issue!

**Comment by squidfunk** [comment]
Released as part of [9.1.15+insiders-4.35.3](https://github.com/squidfunk/mkdocs-material-insiders/releases/tag/9.1.15-insiders-4.35.3)!
